contracts/test/filterMethods.js

```diff
@@ -0,0 +1,167 @@
+        const canonicalBlockHashes = blockHashes.filter(async (blockHash)=> {
+            return (await ethers.provider.getBlock(blockHash)) != null;
+        })
+        expect(currentBlockNumber).gte(startBlockNumber + 2);
+        expect(canonicalBlockHashes.length).gte(currentBlockNumber + 1 - startBlockNumber);
+
+        let lastBlockNumber = 0;
+        for (const blockHash of blockHashes) {
+            const block = await ethers.provider.getBlock(blockHash);
+            if (block != null) {
+                expect(block.number).gte(lastBlockNumber);
+                lastBlockNumber = block.number;
+            }
+        }
+    });
+
+    it("eth_getFilterChanges with filter using address and topics", async function () {
+        // 1. Deploy contract
+        const eventProducer = await ethers.getContractFactory("EventProducer");
+        const contract = await eventProducer.deploy();
+        await contract.deployed();
+
+        // 2. Create a filter with address and topics.
+        //
+        // topics: [signature of event name, $producer, $repeats]
+        const producer = 2;
+        const repeats = 2;
+        const topics = [
+            null,
+            "0x" + ("0".repeat(64) + producer.toString(16)).slice(-64),
+            "0x" + ("0".repeat(64) + repeats.toString(16)).slice(-64),
+        ];
+        const address = contract.address;
+        const filterId = await ethers.provider.send("eth_newFilter", [{address, topics}]);
+        const tx = await contract.produce(producer, repeats);
+        tx.wait();
+
+        // 3. We should get `$repeats` results with order(asc) by id desc
+        const receipt = await ethers.provider.getTransactionReceipt(tx.hash);
+        const logs = await ethers.provider.send("eth_getFilterChanges", [filterId]);
+        expect(receipt.logs.map(log=>log.logIndex)).deep.eq(logs.map(log=>Number(log.logIndex)));
+        expect(logs.length).eq(repeats);
+        logs.reduce((lastLogId, log)=>{
+            expect(Number(log.logIndex)).gt(lastLogId);
+            return Number(log.logIndex);
+        }, -1);
+
+        // 4. Filter should return [] now
+        expect(await ethers.provider.send("eth_getFilterChanges", [filterId])).deep.eq([]);
+    });
+});
+
+describe("eth_getFilterLogs", function () {
+    // NOTE: `eth_getFilterLogs` returns [] for eth_newBlockFilter
+    it("eth_getFilterLogs with BlockFilter", async function () {
+        // 1. Create a block filter
+        const startBlockNumber = await ethers.provider.getBlockNumber();
+        const filterId = await ethers.provider.send("eth_newBlockFilter", []);
+
+        // 2. Deploy contract and make a call
+        const eventProducer = await ethers.getContractFactory("EventProducer");
+        const contract = await eventProducer.deploy();
+        await contract.deployed();
+
+        const producer = 3;
+        const repeats = 3;
+        (await contract.produce(producer, repeats)).wait();
```

Comment:
```suggestion
        await (await contract.produce(producer, repeats)).wait(2);
```

need to wait for 2 blocks? otherwise, I don't think the database has been updated

Reply:
I thought `wait(2)` was no longer needed?

Reply:
Oh yes, I need to wait for 2 blocks. `wait(2)` is good for me. Thanks!

Reply:
> I thought `wait(2)` was no longer needed?

we can get rid of this after this pr https://github.com/nervosnetwork/godwoken-web3/pull/334

most of the time wait(1)(not on chain yet) is enough, but logs require database updates, so need to wait(2)(on chain)